tests/flytekit/unit/core/test_type_engine.py

```diff
@@ -375,6 +399,78 @@ def test_dataclass_transformer():
     assert t.metadata is None
 
 
+def check_element_type_is_int(val):
+    if type(val) == list:
+        for v in val:
+            check_element_type_is_int(v)
+    elif type(val) == dict:
+        for k, v in val.items():
+            check_element_type_is_int(k)
+            check_element_type_is_int(v)
+    else:
+        if type(val) is not str:
+            assert type(val) == int
+
+
+def test_dataclass_int_preserving():
+    ctx = FlyteContext.current_context()
+
+    i = InnerStruct(a=5, b=None, c=[1, 2, 3])
+    tf = DataclassTransformer()
+    lv = tf.to_literal(ctx, i, InnerStruct, tf.get_literal_type(InnerStruct))
+    ot = tf.to_python_value(ctx, lv=lv, expected_python_type=InnerStruct)
+    assert ot.a == 5
```

Comment:
why can't we rely on the equality comparison method generated for the dataclass?

Reply:
Aah we should

Reply:
I've updated it, thank you @eapolinario